### DISCUSSION TITLE
About RPC usage

### DISCUSSION THREAD

**Discussion by shurshilov** [reduxjs/redux-toolkit#4228] [General]
RTK is great for REST. But I have a new experience using RPC. And I came across the fact that RTKQ either can't work with it or it can, but I don't understand how.
I have a universal reading function on my backend. Which accepts model (aka table db) parameter, an filter(or domain), offset (for pagination), and a limit. And the problems here are 2. 
1 I need dynamic types because in one component, for example, I request messages and in the other a list of users with one endpoint. And I don't understand how I can dynamically change the type of RTKQ hook. 
2. The problem is that I have various tags for working with invalidate and other things, but I can sort of generate tags with a function based on input parameters, thereby creating tags based on the model parameter. 
In fact, there was a 3 problem, this is caching. But I solved it using the 
```serializeQueryArgs method: ({EndPointName, queryArgs }) =>`${endpointName}-${query Args?.model} ````
thus just breaking down by model and it works)
Could you describe the solution to these problems? Or is it better to create separate endpoints?
    
![image](https://github.com/reduxjs/redux-toolkit/assets/11828278/ac0fe683-9063-4d1e-8919-2882c411c1a6)

**Comment by shurshilov** [comment]
@phryneas Please answer :)

**Reply by markerikson** [reply]
@shurshilov Please don't ping maintainers directly (and _especially_ not after just a single day).

Don't have a specific answer at this point, other than trying to use RTQK for a general-purpose RPC endpoint like that may not be a good fit.

**Reply by shurshilov** [reply]
Hi, six months have passed, I think there is enough time to raise the topic again))
by the way, I found the code in React Query that does this. (opensource project react admin)
Please take a look, especially at line 67. There the type is passed dynamically.
Is it possible to do the same or similar in RTK query?
![image](https://github.com/user-attachments/assets/c2cdc1b5-2c32-467a-9d78-c2bea911f802)

**Reply by shurshilov** [reply]
Link to code
https://github.com/marmelab/react-admin/blob/37ca014f6bb715f6a97b89ed3516b77baccb270e/packages/ra-core/src/dataProvider/useGetList.ts#L80
https://github.com/marmelab/react-admin/blob/37ca014f6bb715f6a97b89ed3516b77baccb270e/packages/ra-core/src/controller/list/useListController.ts#L96

**Reply by phryneas** [reply]
Honestly... there's no type safety here, at all. 
You call the `useQuery` hook with type arguments that determine the result - you could just as well call a RTK Query  `useQuery` hook and do a `as WhatEverYouWant`. 

Generally: If you have three different ways of calling something and you get three different results, I would make three different endpoints out of it. It doesn't matter that they are the same url, they are completely different stuff.

But beyond that, I can't really tell you a lot. If it's not possible to statically type it, use `unknown` or `any` as the result types. It's not that what you're doing there with React Query has any more type safety than that.
